File: Sources/SwiftDSP/Config/ProcessorConfigTypes.swift
import Foundation

public enum ProcessorType: String, Codable, Sendable {
  case compressor = "Compressor"
  case noiseGate = "NoiseGate"
  case race = "RACE"
}

public struct CompressorParameters: Codable, Sendable, Equatable {
  public var channels: Int
  public var monitorChannels: [Int]?
  public var processChannels: [Int]?
  public var attack: Double
  public var release: Double
  public var threshold: Double
  public var factor: Double
  public var makeupGain: Double?
  public var softClip: Bool?
  public var clipLimit: Double?

  enum CodingKeys: String, CodingKey {
    case channels
    case monitorChannels = "monitor_channels"
    case processChannels = "process_channels"
    case attack, release, threshold, factor
    case makeupGain = "makeup_gain"
    case softClip = "soft_clip"
    case clipLimit = "clip_limit"
  }

  public init(
    channels: Int, monitorChannels: [Int]? = nil, processChannels: [Int]? = nil,
    attack: Double, release: Double, threshold: Double, factor: Double,
    makeupGain: Double? = nil, softClip: Bool? = nil, clipLimit: Double? = nil
  ) {
    self.channels = channels
    self.monitorChannels = monitorChannels
    self.processChannels = processChannels
    self.attack = attack
    self.release = release
    self.threshold = threshold
    self.factor = factor
    self.makeupGain = makeupGain
    self.softClip = softClip
    self.clipLimit = clipLimit
  }

  public func monitorChannelsArray() -> [Int] {
    return monitorChannels ?? []
  }

  public func processChannelsArray() -> [Int] {
    return processChannels ?? []
  }

  public func makeupGainValue() -> Double {
    return makeupGain ?? 0.0
  }

  public func softClipValue() -> Bool {
    return softClip ?? false
  }
}

public struct NoiseGateParameters: Codable, Sendable, Equatable {
  public var channels: Int
  public var monitorChannels: [Int]?
  public var processChannels: [Int]?
  public var attack: Double
  public var release: Double
  public var threshold: Double
  public var attenuation: Double

  enum CodingKeys: String, CodingKey {
    case channels
    case monitorChannels = "monitor_channels"
    case processChannels = "process_channels"
    case attack, release, threshold, attenuation
  }

  public init(
    channels: Int, monitorChannels: [Int]? = nil, processChannels: [Int]? = nil,
    attack: Double, release: Double, threshold: Double, attenuation: Double
  ) {
    self.channels = channels
    self.monitorChannels = monitorChannels
    self.processChannels = processChannels
    self.attack = attack
    self.release = release
    self.threshold = threshold
    self.attenuation = attenuation
  }

  public func monitorChannelsArray() -> [Int] {
    return monitorChannels ?? []
  }

  public func processChannelsArray() -> [Int] {
    return processChannels ?? []
  }
}

public struct RACEParameters: Codable, Sendable, Equatable {
  public var channels: Int
  public var channelA: Int
  public var channelB: Int
  public var delay: Double
  public var subsampleDelay: Bool?
  public var delayUnit: DelayUnit?
  public var attenuation: Double

  enum CodingKeys: String, CodingKey {
    case channels
    case channelA = "channel_a"
    case channelB = "channel_b"
    case delay
    case subsampleDelay = "subsample_delay"
    case delayUnit = "delay_unit"
    case attenuation
  }

  public init(
    channels: Int, channelA: Int, channelB: Int, delay: Double,
    subsampleDelay: Bool? = nil, delayUnit: DelayUnit? = nil, attenuation: Double
  ) {
    self.channels = channels
    self.channelA = channelA
    self.channelB = channelB
    self.delay = delay
    self.subsampleDelay = subsampleDelay
    self.delayUnit = delayUnit
    self.attenuation = attenuation
  }

  public func subsampleDelayValue() -> Bool {
    return subsampleDelay ?? false
  }

  public func delayUnitValue() -> DelayUnit {
    return delayUnit ?? .ms
  }
}

public enum ProcessorConfig: Codable, Sendable, Equatable {
  case compressor(CompressorParameters)
  case noiseGate(NoiseGateParameters)
  case race(RACEParameters)

  enum CodingKeys: String, CodingKey {
    case type, parameters
  }

  public var type: ProcessorType {
    switch self {
    case .compressor: return .compressor
    case .noiseGate: return .noiseGate
    case .race: return .race
    }
  }

  public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    let type = try container.decode(ProcessorType.self, forKey: .type)

    switch type {
    case .compressor:
      let p = try container.decode(CompressorParameters.self, forKey: .parameters)
      self = .compressor(p)
    case .noiseGate:
      let p = try container.decode(NoiseGateParameters.self, forKey: .parameters)
      self = .noiseGate(p)
    case .race:
      let p = try container.decode(RACEParameters.self, forKey: .parameters)
      self = .race(p)
    }
  }

  public func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(type, forKey: .type)

    switch self {
    case .compressor(let p):
      try container.encode(p, forKey: .parameters)
    case .noiseGate(let p):
      try container.encode(p, forKey: .parameters)
    case .race(let p):
      try container.encode(p, forKey: .parameters)
    }
  }
}

extension ProcessorConfig {
  public func validate() throws {
    switch self {
    case .compressor(let p):
      guard p.attack > 0 else {
        throw ConfigError.invalidFilter("Compressor: attack must be > 0, got \(p.attack)")
      }
      guard p.release > 0 else {
        throw ConfigError.invalidFilter("Compressor: release must be > 0, got \(p.release)")
      }
      for ch in p.monitorChannelsArray() {
        guard ch < p.channels else {
          throw ConfigError.invalidFilter(
            "Compressor: monitor channel \(ch) is invalid (max: \(p.channels - 1))")
        }
      }
      for ch in p.processChannelsArray() {
        guard ch < p.channels else {
          throw ConfigError.invalidFilter(
            "Compressor: process channel \(ch) is invalid (max: \(p.channels - 1))")
        }
      }
    case .noiseGate(let p):
      guard p.attack > 0 else {
        throw ConfigError.invalidFilter("NoiseGate: attack must be > 0, got \(p.attack)")
      }
      guard p.release > 0 else {
        throw ConfigError.invalidFilter("NoiseGate: release must be > 0, got \(p.release)")
      }
      for ch in p.monitorChannelsArray() {
        guard ch < p.channels else {
          throw ConfigError.invalidFilter(
            "NoiseGate: monitor channel \(ch) is invalid (max: \(p.channels - 1))")
        }
      }
      for ch in p.processChannelsArray() {
        guard ch < p.channels else {
          throw ConfigError.invalidFilter(
            "NoiseGate: process channel \(ch) is invalid (max: \(p.channels - 1))")
        }
      }
    case .race(let p):
      guard p.attenuation > 0 else {
        throw ConfigError.invalidFilter("RACE: attenuation must be > 0, got \(p.attenuation)")
      }
      guard p.delay > 0 else {
        throw ConfigError.invalidFilter("RACE: delay must be > 0, got \(p.delay)")
      }
      guard p.channelA != p.channelB else {
        throw ConfigError.invalidFilter(
          "RACE: channels A and B must be different, got both \(p.channelA)")
      }
      guard p.channelA < p.channels else {
        throw ConfigError.invalidFilter(
          "RACE: channel A \(p.channelA) is invalid (max: \(p.channels - 1))")
      }
      guard p.channelB < p.channels else {
        throw ConfigError.invalidFilter(
          "RACE: channel B \(p.channelB) is invalid (max: \(p.channels - 1))")
      }
    }
  }
}

File: Sources/CDSP/Config/processor_config_types.c
#include "processor_config_types.h"

#include <stdio.h>
#include <string.h>

const char* processor_type_to_string(processor_type_t type) {
  switch (type) {
    case PROCESSOR_TYPE_COMPRESSOR:
      return "Compressor";
    case PROCESSOR_TYPE_NOISE_GATE:
      return "NoiseGate";
    case PROCESSOR_TYPE_RACE:
      return "RACE";
    default:
      return "Compressor";
  }
}

processor_type_t processor_type_from_string(const char* str) {
  if (!str) return PROCESSOR_TYPE_COMPRESSOR;
  if (strcmp(str, "Compressor") == 0) return PROCESSOR_TYPE_COMPRESSOR;
  if (strcmp(str, "NoiseGate") == 0) return PROCESSOR_TYPE_NOISE_GATE;
  if (strcmp(str, "RACE") == 0) return PROCESSOR_TYPE_RACE;
  return PROCESSOR_TYPE_COMPRESSOR;
}

int processor_config_validate(const processor_config_t* proc,
                              config_error_t* err) {
  if (!proc) return 0;
  switch (proc->type) {
    case PROCESSOR_TYPE_COMPRESSOR: {
      const compressor_parameters_t* p = &proc->parameters.compressor;
      if (p->channels <= 0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "Compressor: channels must be > 0, got %d", p->channels);
        return -1;
      }
      if (p->attack <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "Compressor: attack must be > 0, got %g", p->attack);
        return -1;
      }
      if (p->release <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "Compressor: release must be > 0, got %g", p->release);
        return -1;
      }
      for (size_t i = 0; i < p->monitor_channels_count; i++) {
        if (p->monitor_channels[i] < 0 || p->monitor_channels[i] >= p->channels) {
          config_error_set(
              err, CONFIG_ERR_INVALID_FILTER,
              "Compressor: monitor channel %d is invalid (max: %d)",
              p->monitor_channels[i], p->channels - 1);
          return -1;
        }
      }
      for (size_t i = 0; i < p->process_channels_count; i++) {
        if (p->process_channels[i] < 0 || p->process_channels[i] >= p->channels) {
          config_error_set(
              err, CONFIG_ERR_INVALID_FILTER,
              "Compressor: process channel %d is invalid (max: %d)",
              p->process_channels[i], p->channels - 1);
          return -1;
        }
      }
      break;
    }
    case PROCESSOR_TYPE_NOISE_GATE: {
      const noise_gate_parameters_t* p = &proc->parameters.noise_gate;
      if (p->channels <= 0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "NoiseGate: channels must be > 0, got %d", p->channels);
        return -1;
      }
      if (p->attack <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "NoiseGate: attack must be > 0, got %g", p->attack);
        return -1;
      }
      if (p->release <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "NoiseGate: release must be > 0, got %g", p->release);
        return -1;
      }
      for (size_t i = 0; i < p->monitor_channels_count; i++) {
        if (p->monitor_channels[i] < 0 || p->monitor_channels[i] >= p->channels) {
          config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                           "NoiseGate: monitor channel %d is invalid (max: %d)",
                           p->monitor_channels[i], p->channels - 1);
          return -1;
        }
      }
      for (size_t i = 0; i < p->process_channels_count; i++) {
        if (p->process_channels[i] < 0 || p->process_channels[i] >= p->channels) {
          config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                           "NoiseGate: process channel %d is invalid (max: %d)",
                           p->process_channels[i], p->channels - 1);
          return -1;
        }
      }
      break;
    }
    case PROCESSOR_TYPE_RACE: {
      const race_parameters_t* p = &proc->parameters.race;
      if (p->channels <= 0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "RACE: channels must be > 0, got %d", p->channels);
        return -1;
      }
      if (p->attenuation <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "RACE: attenuation must be > 0, got %g",
                         p->attenuation);
        return -1;
      }
      if (p->delay <= 0.0) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "RACE: delay must be > 0, got %g", p->delay);
        return -1;
      }
      if (p->channel_a == p->channel_b) {
        config_error_set(
            err, CONFIG_ERR_INVALID_FILTER,
            "RACE: channels A and B must be different, got both %d",
            p->channel_a);
        return -1;
      }
      if (p->channel_a < 0 || p->channel_a >= p->channels) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "RACE: channel A %d is invalid (max: %d)",
                         p->channel_a, p->channels - 1);
        return -1;
      }
      if (p->channel_b < 0 || p->channel_b >= p->channels) {
        config_error_set(err, CONFIG_ERR_INVALID_FILTER,
                         "RACE: channel B %d is invalid (max: %d)",
                         p->channel_b, p->channels - 1);
        return -1;
      }
      break;
    }
  }
  return 0;
}
